File: packages/ts/cairn-p2p/src/server/management.ts
// Management API for server-mode peers (spec 10.5, 10.7).
//
// Opt-in REST/JSON HTTP API bound to 127.0.0.1:9090 by default.
// Bearer token authentication with constant-time comparison.

import { createServer, IncomingMessage, ServerResponse, Server } from 'node:http';
import { timingSafeEqual } from 'node:crypto';

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

/** Management API configuration. */
export interface ManagementConfig {
  /** Whether the management API is enabled. Default: false. */
  enabled: boolean;
  /** Bind address. Default: '127.0.0.1'. */
  bindAddress: string;
  /** Port. Default: 9090. */
  port: number;
  /** Bearer token for authentication. */
  authToken: string;
}

/** Default management API configuration. */
export function defaultManagementConfig(): ManagementConfig {
  return {
    enabled: false,
    bindAddress: '127.0.0.1',
    port: 9090,
    authToken: '',
  };
}

// ---------------------------------------------------------------------------
// Response types
// ---------------------------------------------------------------------------

/** Information about a paired peer. */
export interface PeerInfo {
  peerId: string;
  name: string;
  connected: boolean;
  lastSeen: string | null;
}

/** Per-peer store-and-forward queue info. */
export interface QueueInfo {
  peerId: string;
  pendingMessages: number;
  oldestMessageAgeSecs: number | null;
  totalBytes: number;
}

/** Per-peer relay statistics. */
export interface PeerRelayStats {
  peerId: string;
  bytesRelayed: number;
  activeStreams: number;
}

/** Relay statistics overview. */
export interface RelayStats {
  activeConnections: number;
  perPeer: PeerRelayStats[];
}

/** Response for GET /peers. */
export interface PeersResponse {
  peers: PeerInfo[];
}

/** Response for GET /queues. */
export interface QueuesResponse {
  queues: QueueInfo[];
}

/** Response for GET /relay/stats. */
export interface RelayStatsResponse {
  relay: RelayStats;
}

/** Response for GET /health. */
export interface HealthResponse {
  status: 'healthy' | 'degraded';
  uptimeSecs: number;
  connectedPeers: number;
  totalPeers: number;
}

// ---------------------------------------------------------------------------
// Shared state
// ---------------------------------------------------------------------------

/** Shared state accessible by all management API handlers. */
export class ManagementState {
  readonly authTokenBytes: Buffer;
  peers: PeerInfo[] = [];
  queues: QueueInfo[] = [];
  relayStats: RelayStats = { activeConnections: 0, perPeer: [] };
  readonly startedAt: number;

  constructor(authToken: string) {
    this.authTokenBytes = Buffer.from(authToken, 'utf-8');
    this.startedAt = Date.now();
  }
}

// ---------------------------------------------------------------------------
// Authentication
// ---------------------------------------------------------------------------

/**
 * Validate a bearer token using constant-time comparison.
 * Returns true if the token matches.
 */
function validateToken(provided: string, expected: Buffer): boolean {
  const providedBytes = Buffer.from(provided, 'utf-8');
  if (providedBytes.length !== expected.length) {
    // Still do a comparison against expected to avoid timing leak on length.
    timingSafeEqual(expected, expected);
    return false;
  }
  return timingSafeEqual(providedBytes, expected);
}

/**
 * Extract bearer token from Authorization header.
 * Returns the token string or null if not present/malformed.
 */
function extractBearerToken(req: IncomingMessage): string | null {
  const header = req.headers['authorization'];
  if (!header || !header.startsWith('Bearer ')) {
    return null;
  }
  return header.slice(7);
}

// ---------------------------------------------------------------------------
// Route handling
// ---------------------------------------------------------------------------

function sendJson(res: ServerResponse, status: number, data: unknown): void {
  res.writeHead(status, { 'Content-Type': 'application/json' });
  res.end(JSON.stringify(data));
}

function handlePeers(state: ManagementState, res: ServerResponse): void {
  const response: PeersResponse = { peers: state.peers };
  sendJson(res, 200, response);
}

function handleQueues(state: ManagementState, res: ServerResponse): void {
  const response: QueuesResponse = { queues: state.queues };
  sendJson(res, 200, response);
}

function handleRelayStats(state: ManagementState, res: ServerResponse): void {
  const response: RelayStatsResponse = { relay: state.relayStats };
  sendJson(res, 200, response);
}

function handleHealth(state: ManagementState, res: ServerResponse): void {
  const totalPeers = state.peers.length;
  const connectedPeers = state.peers.filter((p) => p.connected).length;
  const uptimeSecs = Math.floor((Date.now() - state.startedAt) / 1000);
  const status = connectedPeers > 0 ? 'healthy' : 'degraded';

  const response: HealthResponse = {
    status,
    uptimeSecs,
    connectedPeers,
    totalPeers,
  };
  sendJson(res, 200, response);
}

function handlePairingQr(res: ServerResponse): void {
  sendJson(res, 503, {
    error: 'pairing QR generation not yet available (pending headless pairing integration)',
  });
}

// ---------------------------------------------------------------------------
// Server
// ---------------------------------------------------------------------------

/**
 * Create and start the management API HTTP server.
 *
 * Returns a handle with `close()` to stop the server.
 *
 * @throws Error if authToken is empty.
 */
export function createManagementServer(
  config: ManagementConfig,
  state: ManagementState,
): ManagementServer {
  if (config.authToken === '') {
    throw new Error('management API auth token is empty');
  }

  // Warn on non-loopback bind address.
  if (config.bindAddress !== '127.0.0.1' && config.bindAddress !== '::1') {
    console.warn(
      `Management API exposed on non-loopback interface ${config.bindAddress} without TLS. This is insecure.`,
    );
  }

  const server = createServer((req, res) => {
    // Authentication.
    const token = extractBearerToken(req);
    if (token === null || !validateToken(token, state.authTokenBytes)) {
      sendJson(res, 401, { error: 'unauthorized' });
      return;
    }

    // Routing.
    const url = req.url ?? '/';
    const method = req.method ?? 'GET';

    if (method !== 'GET') {
      sendJson(res, 405, { error: 'method not allowed' });
      return;
    }

    switch (url) {
      case '/peers':
        handlePeers(state, res);
        break;
      case '/queues':
        handleQueues(state, res);
        break;
      case '/relay/stats':
        handleRelayStats(state, res);
        break;
      case '/health':
        handleHealth(state, res);
        break;
      case '/pairing/qr':
        handlePairingQr(res);
        break;
      default:
        sendJson(res, 404, { error: 'not found' });
        break;
    }
  });

  return new ManagementServer(server, config);
}

/** Handle to a running management API server. */
export class ManagementServer {
  private readonly _server: Server;
  private readonly _config: ManagementConfig;

  constructor(server: Server, config: ManagementConfig) {
    this._server = server;
    this._config = config;
  }

  /** Start listening. Returns a promise that resolves when the server is ready. */
  async start(): Promise<void> {
    return new Promise((resolve) => {
      this._server.listen(this._config.port, this._config.bindAddress, () => {
        resolve();
      });
    });
  }

  /** Stop the server. */
  async close(): Promise<void> {
    return new Promise((resolve, reject) => {
      this._server.close((err) => {
        if (err) reject(err);
        else resolve();
      });
    });
  }

  /** The underlying http.Server (for testing). */
  get httpServer(): Server {
    return this._server;
  }
}

File: packages/go/cairn-p2p/server/management.go
package server

import (
	"crypto/subtle"
	"encoding/json"
	"fmt"
	"log"
	"net"
	"net/http"
	"sync"
	"time"

	cairn "github.com/moukrea/cairn/packages/go/cairn-p2p"
)

// ManagementAPI is a minimal interface for server management.
type ManagementAPI interface {
	PairedPeers() []cairn.PeerID
	QueueDepth(peer cairn.PeerID) int
	RelayStats() RelayStatsInfo
}

// RelayStatsInfo holds relay statistics for the management API.
type RelayStatsInfo struct {
	ActiveRelays  int    `json:"active_relays"`
	TotalRelayed  uint64 `json:"total_relayed"`
	RelayCapacity uint32 `json:"relay_capacity"`
	RelayWilling  bool   `json:"relay_willing"`
}

// ServerConfig holds server-mode configuration deltas from standard node defaults.
type ServerConfig struct {
	StoreForwardEnabled bool
	Headless            bool
	Retention           RetentionConfig
}

// DefaultServerConfig returns server-mode defaults (spec/10 section 10.2).
func DefaultServerConfig() ServerConfig {
	return ServerConfig{
		StoreForwardEnabled: true,
		Headless:            true,
		Retention:           DefaultRetentionConfig(),
	}
}

// ---------------------------------------------------------------------------
// Management API HTTP server
// ---------------------------------------------------------------------------

// ManagementConfig holds management API configuration.
type ManagementConfig struct {
	Enabled     bool   `json:"enabled"`
	BindAddress string `json:"bind_address"`
	Port        int    `json:"port"`
	AuthToken   string `json:"auth_token"`
}

// DefaultManagementConfig returns the default management API configuration.
func DefaultManagementConfig() ManagementConfig {
	return ManagementConfig{
		Enabled:     false,
		BindAddress: "127.0.0.1",
		Port:        9090,
		AuthToken:   "",
	}
}

// PeerInfo describes a paired peer for the management API.
type PeerInfo struct {
	PeerID    string  `json:"peer_id"`
	Name      string  `json:"name"`
	Connected bool    `json:"connected"`
	LastSeen  *string `json:"last_seen"`
}

// QueueInfo describes a per-peer store-and-forward queue.
type QueueInfo struct {
	PeerID              string `json:"peer_id"`
	PendingMessages     int    `json:"pending_messages"`
	OldestMessageAge    *int64 `json:"oldest_message_age_secs"`
	TotalBytes          int64  `json:"total_bytes"`
}

// PeerRelayStats holds per-peer relay statistics.
type PeerRelayStats struct {
	PeerID        string `json:"peer_id"`
	BytesRelayed  uint64 `json:"bytes_relayed"`
	ActiveStreams uint32 `json:"active_streams"`
}

// RelayStatsDetail holds relay statistics overview.
type RelayStatsDetail struct {
	ActiveConnections uint32           `json:"active_connections"`
	PerPeer           []PeerRelayStats `json:"per_peer"`
}

// PeersResponse is the JSON response for GET /peers.
type PeersResponse struct {
	Peers []PeerInfo `json:"peers"`
}

// QueuesResponse is the JSON response for GET /queues.
type QueuesResponse struct {
	Queues []QueueInfo `json:"queues"`
}

// RelayStatsResponse is the JSON response for GET /relay/stats.
type RelayStatsResponse struct {
	Relay RelayStatsDetail `json:"relay"`
}

// HealthResponse is the JSON response for GET /health.
type HealthResponse struct {
	Status         string `json:"status"`
	UptimeSecs     int64  `json:"uptime_secs"`
	ConnectedPeers int    `json:"connected_peers"`
	TotalPeers     int    `json:"total_peers"`
}

// ManagementState holds shared state for the management API handlers.
type ManagementState struct {
	mu         sync.RWMutex
	authToken  []byte
	peers      []PeerInfo
	queues     []QueueInfo
	relayStats RelayStatsDetail
	startedAt  time.Time
}

// NewManagementState creates a new management state with the given auth token.
func NewManagementState(authToken string) *ManagementState {
	return &ManagementState{
		authToken:  []byte(authToken),
		peers:      []PeerInfo{},
		queues:     []QueueInfo{},
		relayStats: RelayStatsDetail{PerPeer: []PeerRelayStats{}},
		startedAt:  time.Now(),
	}
}

// SetPeers updates the peers list.
func (s *ManagementState) SetPeers(peers []PeerInfo) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.peers = peers
}

// SetQueues updates the queues list.
func (s *ManagementState) SetQueues(queues []QueueInfo) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.queues = queues
}

// SetRelayStats updates the relay statistics.
func (s *ManagementState) SetRelayStats(stats RelayStatsDetail) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.relayStats = stats
}

// ManagementServer is the HTTP server for the management API.
type ManagementServer struct {
	config ManagementConfig
	state  *ManagementState
	server *http.Server
}

// NewManagementServer creates a new management API server.
// Returns an error if the auth token is empty.
func NewManagementServer(config ManagementConfig, state *ManagementState) (*ManagementServer, error) {
	if config.AuthToken == "" {
		return nil, fmt.Errorf("management API auth token is empty")
	}

	// Warn on non-loopback bind address.
	ip := net.ParseIP(config.BindAddress)
	if ip != nil && !ip.IsLoopback() {
		log.Printf("WARNING: Management API exposed on non-loopback interface %s without TLS. This is insecure.", config.BindAddress)
	}

	mux := http.NewServeMux()
	ms := &ManagementServer{
		config: config,
		state:  state,
	}

	mux.HandleFunc("/peers", ms.withAuth(ms.handlePeers))
	mux.HandleFunc("/queues", ms.withAuth(ms.handleQueues))
	mux.HandleFunc("/relay/stats", ms.withAuth(ms.handleRelayStats))
	mux.HandleFunc("/health", ms.withAuth(ms.handleHealth))
	mux.HandleFunc("/pairing/qr", ms.withAuth(ms.handlePairingQR))

	ms.server = &http.Server{
		Addr:    fmt.Sprintf("%s:%d", config.BindAddress, config.Port),
		Handler: mux,
	}

	return ms, nil
}

// withAuth wraps a handler with bearer token authentication.
func (ms *ManagementServer) withAuth(handler http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		auth := r.Header.Get("Authorization")
		const prefix = "Bearer "
		if len(auth) <= len(prefix) || auth[:len(prefix)] != prefix {
			writeJSON(w, http.StatusUnauthorized, map[string]string{"error": "unauthorized"})
			return
		}

		provided := []byte(auth[len(prefix):])
		expected := ms.state.authToken

		if len(provided) != len(expected) || subtle.ConstantTimeCompare(provided, expected) != 1 {
			writeJSON(w, http.StatusUnauthorized, map[string]string{"error": "unauthorized"})
			return
		}

		handler(w, r)
	}
}

func writeJSON(w http.ResponseWriter, status int, data interface{}) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(data) //nolint:errcheck
}

func (ms *ManagementServer) handlePeers(w http.ResponseWriter, r *http.Request) {
	ms.state.mu.RLock()
	peers := ms.state.peers
	ms.state.mu.RUnlock()
	writeJSON(w, http.StatusOK, PeersResponse{Peers: peers})
}

func (ms *ManagementServer) handleQueues(w http.ResponseWriter, r *http.Request) {
	ms.state.mu.RLock()
	queues := ms.state.queues
	ms.state.mu.RUnlock()
	writeJSON(w, http.StatusOK, QueuesResponse{Queues: queues})
}

func (ms *ManagementServer) handleRelayStats(w http.ResponseWriter, r *http.Request) {
	ms.state.mu.RLock()
	stats := ms.state.relayStats
	ms.state.mu.RUnlock()
	writeJSON(w, http.StatusOK, RelayStatsResponse{Relay: stats})
}

func (ms *ManagementServer) handleHealth(w http.ResponseWriter, r *http.Request) {
	ms.state.mu.RLock()
	peers := ms.state.peers
	ms.state.mu.RUnlock()

	totalPeers := len(peers)
	connectedPeers := 0
	for _, p := range peers {
		if p.Connected {
			connectedPeers++
		}
	}

	uptimeSecs := int64(time.Since(ms.state.startedAt).Seconds())

	status := "degraded"
	if connectedPeers > 0 {
		status = "healthy"
	}

	writeJSON(w, http.StatusOK, HealthResponse{
		Status:         status,
		UptimeSecs:     uptimeSecs,
		ConnectedPeers: connectedPeers,
		TotalPeers:     totalPeers,
	})
}

func (ms *ManagementServer) handlePairingQR(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, http.StatusServiceUnavailable, map[string]string{
		"error": "pairing QR generation not yet available (pending headless pairing integration)",
	})
}

// Start begins listening and serving. Blocks until the server stops.
func (ms *ManagementServer) Start() error {
	return ms.server.ListenAndServe()
}

// Close stops the management server.
func (ms *ManagementServer) Close() error {
	return ms.server.Close()
}

// Handler returns the http.Handler for testing purposes.
func (ms *ManagementServer) Handler() http.Handler {
	return ms.server.Handler
}
